### ISSUE TITLE
Your Name: module name, number of weeks in this module

### ISSUE THREAD

**Issue by alihaidermalik20** [lab-brussels-1/home#356]
> [my fork of module repository]()

## Learning Objectives

<details>
<summary>Priorities: 🥚, 🐣, 🐥, 🐔 (click to learn more)</summary>
<br>

There is a lot to learn in this repository. If you can't master all the material
at once, that's expected! Anything you don't master now will always be waiting
for you to review when you need it. These 4 emoji's will help you prioritize
your study time and to measure your progress:

- 🥚: Understanding this material is required, it covers the base skills you'll
  need to move on. You do not need to finish all of them but should feel
  comfortable that you could with enough time.
- 🐣: You have started all of these exercises and feel you could complete them
  all if you just had more time. It may not be easy for you but with effort you
  can make it through.
- 🐥: You have studied the examples and started some exercises if you had time.
  You should have a big-picture understanding of these concepts/skills, but may
  not be confident completing the exercises.
- 🐔: These concepts or skills are not necessary but are related to this module.
  If you are finished with 🥚, 🐣 and 🐥 you can use the 🐔 exercises to push
  yourself without getting distracted from the module's main objectives.

---

</details>

### [1. Remix](./1-remix/)

Practice studying and remixing other people's solutions to coding challenges. Create your own solutions by mixing and matching pieces from other people's code.

- 🐣 **Reconstructing**: You can reconstruct a variety of solutions to the same challenge when they are presented as parsons problems.
- 🐣 **Analyzing**: You can analyze a function written at your level. This includes:
  - [x] **Behavior**: Write documentation, test cases and use cases to describe the function's behavior.
  - [x] **Strategy**: Describe the function's strategy using plain english.
  - [x] **Implementation**: List the language features in a function and explain how each one is used.
  - [x] **Small Changes**: You can think of 2+ changes to the function's _implementation_ that would not change it's _strategy_.
- 🐣 **Remixing** You can analyze several solutions to the same challenge then ...
  - [x] **Write**: Your own solution by remixing the ones you studied.
  - [x] **Explain**: How the other solutions inspired yours; what ideas did you take from them? what ideas did you _not_ take?
  - [x] **Analyze**: Complete a write-up your own solution as though someone else wrote it.

### [2. Write](./2-write/)

- 🐣 **Function Design**:
  - [ ] **Writing Tests**: Given a working function, you can write passing test cases to describe it's behavior.
  - [ ] **Writing Functions**: You can design multiple solutions to the same code challenge, keeping notes about different experiments you tried along the way.
- [ ] 🐣 **Generating Documentation**: You can write a JSDoc comment for your solutions and run a script to generate markdown documentation.
- [ ] 🐣 **Fuzz Testing**: You write solutions that pass randomly generated test cases.
- 🐣 **Test Driven Development**: You can solve open-ended, ambiguous coding challenges at your level:
  - [ ] **Reading Docs**: You can understand what the function is supposed to do by reading it's JSDoc description.
  - [ ] **Writing Tests**: You can write test cases _before_ there is a function to test.
  - [ ] **Writing Functions**: You can write one function that passes the test cases you have prepared (even if it's just 1 test case!).
  - [ ] **Refactoring**: You can improve your function's implementation _without_ failing any test cases that were passing.
  - [ ] **Iterative Development**: You can repeat the TDD process until you are satisfied with your test cases and solution.
- [ ] 🐔 **Code Golf**: Write your solutions with the fewest characters possible! This won't help you write readable code, but it will make you think deeply about JS, your strategy and implementation.

### [3. Review](./3-review/)

- 🐣 **Continuous Integration**: You can check your code's quality _before_ pushing so your CI checks all pass.
  - [ ] **Formatting**
  - [ ] **Linting**
  - [ ] **Testing**
- [ ] 🐣 **Code Review**: You can use a checklist to give a thorough, positive and constructive review of your classmates' solutions.
- [ ] 🐔 **Code Coverage**: You can explain what code coverage is, why it's important, and can write unit tests with 100% code coverage.

---

**Comment by alihaidermalik20** [comment]
Week 1

- [ ] I have pushed my progress to [my fork of exercises repo](https://github.com/alihaidermalik20/solution-write-ups)

## I Need Help With:

## What went well?

- basic testing with Jest
- going through Jest documentation to find out which tests i can use and why some fail even though they should have passed. Such as ThrowError() should be checked against a function unlike toEqual()

## What went less well?

- JSDoc etc. Still need to practice more of the standard format for documentation of functions

## Lessons Learned

- learning about Jest and that there are many other testing programs available

## Sunday Prep Work

**Comment by danielhalasz** [comment]
> * ThrowError() should be checked against a function unlike toEqual()

yes, that is a tricky one :) 

> JSDoc etc. Still need to practice more of the standard format for documentation of functions

there is a video about it by Mr Traversy: https://www.youtube.com/watch?v=YK-GurROGIg

**Comment by alihaidermalik20** [comment]
Week 2

- [x] I have pushed my progress to [my fork of exercises repo](https://github.com/alihaidermalik20/docker-example-bsi)

## I Need Help With:

## What went well?

- 505 tests passed for the doctor nurse example
- created a function with proper styling/linting and 2 space indentation. JS styling is somewhat different to C or python
- how to use APIs and async await. working with axios. went through some node documentation to see how node repl works and other node functions that can be used with require such as fs and fs.writeFile Sync or non Sync

## What went less well?

## Lessons Learned

- testing for each and every case with a loop and not just one test for each case

## Sunday Prep Work

**Comment by danielhalasz** [comment]
> * 505 tests passed for the doctor nurse example

awesome job Ali!!

**Comment by danielhalasz** [comment]
> * created a function with proper styling/linting and 2 space indentation. JS styling is somewhat different to C or python

yes, JS is more forgiving in terms of indentation..but it is good to do it properly anyway (and Prettier and other tools can make this automatic), it helps finding bugs and learning more strict languages

**Comment by danielhalasz** [comment]
> * how to use APIs and async await. working with axios. went through some node documentation to see how node repl works and other node functions that can be used with require such as fs and fs.writeFile Sync or non Sync

it is great that you looked into this, HYF covered how to write to the file system before, but now it is more focused on front-end.. 

what have you used axios for? which API did you connect to? what requests did you make and with what tool? or so far it has been more theoretical?

**Comment by danielhalasz** [comment]
> * testing for each and every case with a loop and not just one test for each case

yes, this was a cool trick shown by Thibault

**Comment by danielhalasz** [comment]
you can also have a look at how Maria solved it:
https://github.com/MMikhailova/docker-example-bsi/blob/master/doctor-solution.test.js

**Comment by alihaidermalik20** [comment]
> > * how to use APIs and async await. working with axios. went through some node documentation to see how node repl works and other node functions that can be used with require such as fs and fs.writeFile Sync or non Sync
> 
> it is great that you looked into this, HYF covered how to write to the file system before, but now it is more focused on front-end..
> 
> what have you used axios for? which API did you connect to? what requests did you make and with what tool? or so far it has been more theoretical?

i used axios to make it easier to use async and await for get requests only. Get a data response from an api and enter select the image from the response data and append it on a website if an image exists in that data.

**Comment by alihaidermalik20** [comment]
> you can also have a look at how Maria solved it: https://github.com/MMikhailova/docker-example-bsi/blob/master/doctor-solution.test.js

oh wow! that is a very different solution

**Comment by alihaidermalik20** [comment]
Week 3

- [x] I have pushed my progress to [my fork of exercises repo](https://github.com/lab-brussels-1/group-6-practice-code-review)

## I Need Help With:

## What went well?

- revisiting github and git collaboration
- took a few days break/holiday from my laptop to avoid burnout
- learned how to use express and ejs templating which was very cool how you can make a template out of header, footer, navbar etc. and use them for every page by just translating js into html. 
- It's nice to go through more advanced lessons now as i forget a lot of it, so when we cover it at HYF, i'll be revisiting it and cementing it to my memory. 

## What went less well?

-  sort numbers exercises is a bit tricky as the sort method is confusing even with the extensive MDN docs.

## Lessons Learned

- sometimes just reading the docs isn't enough and you have to try different things on your own machine to fully understand

## Sunday Prep Work
